### PR TITLE
Document new TestSystem functionality

### PR DESCRIPTION
**Pull request by nvoxland**
## Overview

Start documenting the new TestSystem functionality. It is alpha-level APIs, but ready to start getting feedback.

Including it in a new DEVELOPING.md file which can eventually be the starting page for all developer-level docs we wish to include in the repo.

### CODE REVIEW

**Comment by kataggart** [comment]
@kevin-atx can you please review? Thanks

**Comment by nvoxland** [comment]
Moving this along for others to take a look at

**Review by MalloD12 — APPROVED** [review]
Approved.

In my opinion TestSystem documentation is good enough for now, if we think any other detail should be added we will update this document.

**Review by filipelautert — APPROVED** [review]
That's useful!

File: DEVELOPING.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
+# Developing Liquibase
+
+## Overview
+
+Liquibase is implemented in Java, and uses Maven as it's build system.
```
Comment (XDelphiGrl) [inline DEVELOPING.md:5]: `it's` should be `its`.

File: DEVELOPING.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+The base TestSystem class is designed to work for more than just Databases, so if you know you are connecting to a database you will want to cast it to a DatabaseTestSystem in order to call methods like `getConnection()` for use in connecting to it. 
+For example:
+```java
+@Shared
+private DatabaseTestSystem mysql = 
+    (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("mysql")
+```
+
+When TestSystems need to be started, Liquibase generally relies on Docker instances. For time when docker images are not available or are unneeded, they can also connect to an external database via a `url` setting.
```
Comment (XDelphiGrl) [inline DEVELOPING.md:39]: Minor edit for the sentence at line 39 (delete the "For time")

When docker images are not available or are unneeded, they can also connect to an external database via a `url` setting.

File: DEVELOPING.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+For example:
+```java
+@Shared
+private DatabaseTestSystem mysql = 
+    (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("mysql")
+```
+
+When TestSystems need to be started, Liquibase generally relies on Docker instances. For time when docker images are not available or are unneeded, they can also connect to an external database via a `url` setting. 
+
+When using docker instances, if we want to add support for a new database we want to test against, we need to create a new test class for which we will need to extend from `DatabaseTestSystem` and override
```
Comment (XDelphiGrl) [inline DEVELOPING.md:41]: Minor edit to line 41 to tighten up the wording a bit:

"... we need to create a new test class extended from `DatabaseTestSystem` and override ..."

File: DEVELOPING.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+<dependency>
+    <groupId>org.firebirdsql</groupId>
+    <artifactId>firebird-testcontainers-java</artifactId>
+    <version>1.2.0</version>
+</dependency>
+```
+
+#### Controlling which systems are tested against
+
+Even though systems are able to be automatically brought up, tests against databases are slow. Therefore, exactly which test systems should be tested against is controllable by you.
```
Comment (XDelphiGrl) [inline DEVELOPING.md:74]: Edits for line 74 to tighten up language:

* The test framework is designed to automatically launch database instances of any type. However, database tests can be slow, so Liquibase defaults to testing in-memory databases (h2, hsqldb, sqlite). Any tests requiring other database platforms will be automatically marked as "ignored".

File: DEVELOPING.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+By default, Liquibase will only test against in-memory databases (h2, hsqldb, sqlite) because they are fast. Any tests which ask for other systems will automatically be marked as "ignored".
+
+To change which tests run and which are ignored, set the `liquibase.sdk.testSystem.test` attribute to be a comma-separated list. This can be most easily set in your liquibase.sdk.local.yml file (see below).
```
Comment (XDelphiGrl) [inline DEVELOPING.md:77]: Edit to line 77 to simply state the recommended method to change which tests run.

"The `liquibase.sdk.testSystem.test` attribute is a comma-separated list specified in the liquibase.sdk.local.yml (see below). "

File: DEVELOPING.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+The `liquibase.sdk.testSystem.test` setting can also specify specific settings for the test systems you wish to test against. 
+For example, if you have `liquibase.sdk.testSystem.test=h2?version=1.4.200,mysql`, when your tests call `getTestSystem("h2")` it will get an h2 version 1.4.200 system. You can also specify profiles (see below) as well, like `liquibase.sdk.testSystem.test=mssql:case-sensitive,mysql`
```
Comment (XDelphiGrl) [inline DEVELOPING.md:80]: This is pretty neat!

File: DEVELOPING.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+The `liquibase-extension-testing.jar` and corresponding `liquibase-extension-testing-deps.jar` files add a `liquibase sdk system` command group to the Liquibase CLI which allow you to start and stop the same containers used by your test.
+By adding those jars to your LIQUIBASE_HOME/lib directory, you can run `liquibase sdk system up --name mysql` to bring up a mysql system, where the `name` argument is the name of the test system you want to start.
+
+To shut down the instances started this way, run `liquibase sdk system down --name mysql`
+
+#### Local configuration
+
+The liquibase.sdk.yaml file contains the default configuration, but there are times you may want to change the default behavior. 
+By creating a `liquibase.sdk.local.yml` file in the [same directory](liquibase-extension-testing/src/main/resources) as the liquibase.sdk.yaml file, you can add just the settings you want to override.
+NOTE: gitignore is set to not allow the file to be committed because it's designed to be specific to you.
```
Comment (XDelphiGrl) [inline DEVELOPING.md:94]: Edit at line 94:

"NOTE: The `liquibase.sdk.local.yml` is specific to your development environment and gitignore is configured to disallow the  file from being committed."

File: DEVELOPING.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+```yaml
+  h2:
+    url: jdbc:h2:mem:${catalog};DB_CLOSE_DELAY=-1
+    profiles:
+      "1.x":
+        version: 1.4.200
+      "2.x":
+        version: 2.0.206
+```
+which makes it easier for users who just care about the top-level version.
```
Comment (XDelphiGrl) [inline DEVELOPING.md:128]: Edit at line 128:

"The advantage of profiles is it allows users to specify the exact version of h2 relevant for their testing."

**Review by XDelphiGrl — CHANGES_REQUESTED** [review]
@nvoxland, hi - I've made some phrasing suggestions. 

CHANGES REQUESTED

File: DEVELOPING.md (hunk 4 above)
Comment (MalloD12) [reply DEVELOPING.md]: Done.

File: DEVELOPING.md (hunk 5 above)
Comment (MalloD12) [reply DEVELOPING.md]: Done.

File: DEVELOPING.md (hunk 7 above)
Comment (MalloD12) [reply DEVELOPING.md]: Done.

File: DEVELOPING.md (hunk 8 above)
Comment (MalloD12) [reply DEVELOPING.md]: Done.

**Review by XDelphiGrl — APPROVED** [review]
This PR documents how to use the Liquibase test system.

APPROVED

**Review by suryaaki2 — APPROVED** [review]

**Pull request merged**